### PR TITLE
fixes 17648: add tags and description for tableau published data source

### PR DESCRIPTION
**Pull request by nicor88**
### Describe your changes:

Fixes https://github.com/open-metadata/OpenMetadata/issues/17648

In addition  introduce TableauDataModelTag to deal with tags in published data sources in tableau.

#### Notes
* I added the support to fetch hasExtracts for data models, this is quite relevant to know, I just don't know how to show this field in the UI - can be done in the next iteration

#### TODO
- [ ] add unit tests
- [ ] configure dev environment properly

#
### Type of change:

- [ ] Bug fix
- [x] Improvement
- [ ] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/developers/contribute) document.
- [x] My PR title is `Fixes <issue-number>: <short explanation>`
- [x] I have commented on my code, particularly in hard-to-understand areas. 
- [ ] For JSON Schema changes: I updated the migration scripts or explained why it is not needed.

### CODE REVIEW

File: ingestion/src/metadata/ingestion/source/dashboard/tableau/queries.py [hunk 1]
```diff
@@ -23,9 +23,15 @@
     nodes {{
       id
       name
+      hasExtracts
       upstreamDatasources{{
         id
         name
+        description
+        hasExtracts
```
Comment (OnkarVO7) [inline ingestion/src/metadata/ingestion/source/dashboard/tableau/queries.py:31]: is `hasExtracts` being used anywhere? If not we can remove the same

File: ingestion/src/metadata/ingestion/source/dashboard/tableau/models.py [hunk 2]
```diff
@@ -121,6 +128,9 @@ class UpstreamTable(BaseModel):
 class DataSource(BaseModel):
     id: str
     name: Optional[str] = None
+    description: Optional[str] = None
+    hasExtracts: Optional[bool] = None
+    tags: Optional[List[TableauDataModelTag]] = []
```
Comment (OnkarVO7) [inline ingestion/src/metadata/ingestion/source/dashboard/tableau/models.py:133]: We should assign `None` here as the default and handle the empty list on the metadata.py side

File: ingestion/src/metadata/ingestion/source/dashboard/tableau/metadata.py [hunk 3]
```diff
@@ -160,8 +174,17 @@ def yield_tags(
                 for elem in container:
                     tags.update(elem.tags)
 
+            _tags = {tag.label for tag in tags}
+            logger.debug(f"Tags: {_tags}")
```
Comment (OnkarVO7) [inline ingestion/src/metadata/ingestion/source/dashboard/tableau/metadata.py:178]: We can remove the debug statements from here.

File: ingestion/src/metadata/ingestion/source/dashboard/tableau/metadata.py [hunk 4]
```diff
@@ -160,8 +174,17 @@ def yield_tags(
...
+
+            # we fetch tags from the data models upstream datasources
+            _data_models_tags = self._get_data_models_tags(dashboard_details.dataModels)
+            logger.debug(f"Data Models Tags: {_data_models_tags}")
```
Comment (OnkarVO7) [inline ingestion/src/metadata/ingestion/source/dashboard/tableau/metadata.py:182]: Same here we can remove the debug statements

File: ingestion/src/metadata/ingestion/source/dashboard/tableau/metadata.py [hunk 5]
```diff
@@ -148,6 +148,20 @@ def get_owner_ref(
...
 
+    @staticmethod
+    def _get_data_models_tags(dataModels: [DataSource]) -> Set[str]:
+        """
+        Get the tags from the data model
+        """
+        tags = set()
+        for data_model in dataModels:
+            # we fetch the tags only from the upstream data sources
+            for upstream_source in data_model.upstreamDatasources:
```
Comment (OnkarVO7) [inline ingestion/src/metadata/ingestion/source/dashboard/tableau/metadata.py:159]: lets add `data_model.upstreamDatasources or []` to handle `None` cases

File: ingestion/src/metadata/ingestion/source/dashboard/tableau/metadata.py [hunk 6]
```diff
@@ -148,6 +148,20 @@ def get_owner_ref(
             logger.warning(f"Could not fetch owner data due to {err}")
         return None
 
+    @staticmethod
+    def _get_data_models_tags(dataModels: [DataSource]) -> Set[str]:
```
Comment (OnkarVO7) [inline ingestion/src/metadata/ingestion/source/dashboard/tableau/metadata.py:152]: Lets add a try/except to this method

**Review by OnkarVO7 — CHANGES_REQUESTED** [review]
HI @nicor88 thanks for contribution 🚀 and for raising the PR. Left some minor comments overall it looks good

File: ingestion/src/metadata/ingestion/source/dashboard/tableau/models.py (hunk 2 above)
Comment (nicor88) [reply ingestion/src/metadata/ingestion/source/dashboard/tableau/models.py:133]: I can set None, but I don't fully agree that should be None here.
If you look into [TableauDashboard](https://github.com/open-metadata/OpenMetadata/blob/1924f44b50765e275037868455e5c693544596b3/ingestion/src/metadata/ingestion/source/dashboard/tableau/models.py#L170), we use a list of empty tags.

When dealing with list, I generally prefer to use empty list, vs None objects.

File: ingestion/src/metadata/ingestion/source/dashboard/tableau/queries.py (hunk 1 above)
Comment (nicor88) [reply ingestion/src/metadata/ingestion/source/dashboard/tableau/queries.py:31]: As I wrote in the original PR my idea was to add it, and then eventually in a future iteration consider to see how to see if a data-source has extracts or not, but as I believe that this is another topic I will simply remove it for now:)

File: ingestion/src/metadata/ingestion/source/dashboard/tableau/models.py (hunk 2 above)
Comment (OnkarVO7) [reply ingestion/src/metadata/ingestion/source/dashboard/tableau/models.py:132]: Makes sense it should be fine

**Review by OnkarVO7 — APPROVED** [review]

**Pull request merged**